File: app/src/main/java/hr/ml/moviesdiscover/repository/MoviesRepository.java
package hr.ml.moviesdiscover.repository;

import java.util.List;

import hr.ml.moviesdiscover.BuildConfig;
import hr.ml.moviesdiscover.model.Movie;
import hr.ml.moviesdiscover.rest.TmdbApi;
import hr.ml.moviesdiscover.rest.TmdbRetrofitService;
import hr.ml.moviesdiscover.rest.model.DiscoveredMoviesFromRetrofit;
import hr.ml.moviesdiscover.rest.model.MovieFromRetrofit;
import hr.ml.moviesdiscover.util.MovieMapper;
import hr.ml.moviesdiscover.viewmodel.IMovieDiscoverViewModel;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MoviesRepository implements IMoviesRepository, Callback<DiscoveredMoviesFromRetrofit> {

    private TmdbApi tmdbApi;

    private IMovieDiscoverViewModel viewModel;

    public MoviesRepository(IMovieDiscoverViewModel viewModel) {
        this.viewModel = viewModel;
        tmdbApi = TmdbRetrofitService.Companion.getInstance().create(TmdbApi.class);
    }

    public void requestMovies() {
        Call<DiscoveredMoviesFromRetrofit> discoveredMoviesFromApi = tmdbApi
                .getDiscoveredMovies(BuildConfig.API_KEY, "popularity.desc",
                        false, false, 1);

        discoveredMoviesFromApi.enqueue(this);
    }

    @Override
    public void onResponse(Call<DiscoveredMoviesFromRetrofit> call,
                           Response<DiscoveredMoviesFromRetrofit> response) {
        if (call.isExecuted()) {
            if(response.body() != null) {
                List<MovieFromRetrofit> moviesFromRetrofit = response.body().getResults();
                List<Movie> movies = MovieMapper.Companion
                        .fromMovieOverviewToMovie(moviesFromRetrofit);
                viewModel.fetchMovies(movies);
            }
        } else if (call.isCanceled()) {
            viewModel.fetchMovies(null);
        }
    }

    @Override
    public void onFailure(Call<DiscoveredMoviesFromRetrofit> call, Throwable t) {
        viewModel.fetchMovies(null);
    }
}

File: app/src/main/kotlin/hr/ml/moviesdiscover/repository/MoviesRepository.kt
package hr.ml.moviesdiscover.repository

import hr.ml.moviesdiscover.BuildConfig
import hr.ml.moviesdiscover.rest.TmdbApi
import hr.ml.moviesdiscover.rest.TmdbRetrofitService
import hr.ml.moviesdiscover.rest.model.DiscoveredMoviesFromRetrofit
import hr.ml.moviesdiscover.util.MovieMapper
import hr.ml.moviesdiscover.viewmodel.IMovieDiscoverViewModel
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MoviesRepository(private val viewModel: IMovieDiscoverViewModel) :
        IMoviesRepository, Callback<DiscoveredMoviesFromRetrofit> {

    private val tmdbApi: TmdbApi = TmdbRetrofitService.getInstance().create(TmdbApi::class.java)

    override fun requestMovies() {
        val restCall = tmdbApi.getDiscoveredMovies(BuildConfig.API_KEY,
                "popularity.desc", includeAdult = false, includeVideo = false, page = 1)

        restCall.enqueue(this)
    }

    override fun onResponse(call: Call<DiscoveredMoviesFromRetrofit>,
                            response: Response<DiscoveredMoviesFromRetrofit>) {

        if (response.isSuccessful) {
            val moviesFromRetrofit = response.body()?.results
            val movies = MovieMapper.fromMovieOverviewToMovie(moviesFromRetrofit)
            viewModel.fetchMovies(movies)
        } else viewModel.fetchMovies(emptyList())
    }

    override fun onFailure(call: Call<DiscoveredMoviesFromRetrofit>, t: Throwable) {
        viewModel.fetchMovies(emptyList())
    }
}